File: Medium/2026-03-19-3212-Count-Submatrices-With-Equal-Frequency-of-X-and-Y/solution.java
class Solution {
    public int countSubmatrices(String[][] grid) {
        int m = grid.length, n = grid[0].length;
        int count = 0;

        for (int top = 0; top < m; top++) {
            int[] freq = new int[2]; // freq[0] for 'X', freq[1] for 'Y'
            for (int bottom = top; bottom < m; bottom++) {
                for (int j = 0; j < n; j++) {
                    if (grid[bottom][j].equals("X")) {
                        freq[0]++;
                    } else if (grid[bottom][j].equals("Y")) {
                        freq[1]++;
                    }
                }

                for (int j = 0; j < n; j++) {
                    if (grid[bottom][j].equals("X") || grid[bottom][j].equals("Y")) {
                        int xCount = 0, yCount = 0;
                        for (int k = j; k < n; k++) {
                            if (grid[bottom][k].equals("X")) {
                                xCount++;
                            } else if (grid[bottom][k].equals("Y")) {
                                yCount++;
                            }
                            if (xCount > 0 && xCount == yCount) {
                                count++;
                            }
                            if (grid[bottom][k].equals(".")) {
                                break;
                            }
                        }
                    }
                }
            }
        }
        return count;
    }
}

File: Medium/2026-03-19-3212-Count-Submatrices-With-Equal-Frequency-of-X-and-Y/solution.js
class Solution {
    countSubmatrices(grid) {
        const m = grid.length;
        const n = grid[0].length;
        let count = 0;

        for (let top = 0; top < m; top++) {
            const freq = new Map();

            for (let bottom = top; bottom < m; bottom++) {
                let xCount = 0;
                let yCount = 0;

                for (let col = 0; col < n; col++) {
                    if (grid[bottom][col] === 'X') {
                        xCount++;
                    } else if (grid[bottom][col] === 'Y') {
                        yCount++;
                    }

                    if (bottom === top) {
                        if (grid[bottom][col] === 'X' || grid[bottom][col] === 'Y') {
                            freq.set(col, (freq.get(col) || 0) + 1);
                        }
                    } else {
                        if (grid[bottom][col] === 'X' || grid[bottom][col] === 'Y') {
                            freq.set(col, (freq.get(col) || 0) + 1);
                        }
                    }

                    if (xCount > 0 && xCount === yCount) {
                        count++;
                    }
                }
            }
        }

        return count;
    }
}